### DISCUSSION TITLE
Suggestions on custom processing of (audio) assets

### DISCUSSION THREAD

**Discussion by ChappIO** [directus/directus#17349] [Archive]
I am working on a personal project where I would love to use Directus as my backend.

However, there is one feature I would have to built myself:

I will be sharing high quality .wav files and would like to generate low-quality .mp3 files on-demand just like the thumbnail generation system.

Is there a way for me to expand the media generation system with new file types so I can add support for .wav to .mp3 conversion while benefiting from the build-in caching mechanism? Or will I be better off implementing this entire functionality in a custom endpoint?

**Comment by ChappIO** [comment]
For context, I know how to generate the .mp3 file from the .wav and what codec/quality parameters to use. But if I would be able to hook into the system the thumbnail generator uses I wouldn't have to build the on-demand conversion and caching.

**Comment by rijkvanzanten** [comment]
There's no way to extend the existing `/assets` endpoint. You could build this as a custom endpoint though! 

(Also let me know if the way you're thinking of making this work would scale enough to be part of core!)

**Reply by ChappIO** [reply]
I mean ideally I would be able to register a mimetype into the /assets endpoint where I can receive custom settings (access to the query params). Then all I really have to do this provide a function that transforms the binary to a different format and let Directus handle the rest.

That would open so many door for asset management in Directus.

Now I am planning on making a (hidden) collection that serves as a cache and O2M relationship to a file. It'll be a lot more effort to make it perform well. But I will let you know if I'll go this route.

**Reply by rijkvanzanten** [reply]
Good idea! Would you imagine it would effectively work as a hook? Would it tap into the existing `format` query parameter that's used to transform?

**Reply by ChappIO** [reply]
You know... I might try to implement this in a fork and create a PR next month.

Reading the code here, I think a hook would be very feasible: https://github.com/directus/directus/blob/1bfd46044928a8192fbef7c1a507899b31642abf/api/src/services/assets.ts#L179

**Reply by ChappIO** [reply]
@rijkvanzanten I think it should work based on the source mime type of the original asset.

So I could register a transformer that can deal with audio/x-wav and it would get access to the query parameters so you can specify your own transformation parameters.

In my case I would like to know the target format and quality index required by ffmpeg so I would call:

```
GET /assets/<id>?download&format=mp3&quality=5
```

Directus would already know <id> to be of type audio/x-wav because I uploaded a .wav file.

I could register my hook (pseudo code):
```
const assetService = getAssetService();

assetService.registerTransformer('audio/x-wav', (source, parameters) => {
  const newFileName = source.fileName.replaceWavWithMp3();
  transform(source, {
    quality: parameters.quality || 5,
    target: newFileName
});
return { newFileName };
});
```

Something along those lines.

**Reply by ChappIO** [reply]
I am thinking about implementing this as a filter event & hook.
Right now, is there a reason you'd see why that wouldn't work?

I would introduce the `asset.transform` filter event and launch that from the `AssetService` if the target format was not found yet.
